python/pyspark/sql/connect/client/reattach.py

```diff
@@ -0,0 +1,225 @@
+                        )
+
+                    if self._current is None:
+                        try:
+                            self._current = next(self._iterator)
+                        except StopIteration:
+                            pass
+
+                    has_next = self._current is not None
+
+                    # Graceful reattach:
+                    # If iterator ended, but there was no ResponseComplete, it means that
+                    # there is more, and we need to reattach. While ResponseComplete didn't
+                    # arrive, we keep reattaching.
+                    first_loop = True
+                    if not self._result_complete and not has_next:
+                        while not has_next or first_loop:
+                            self._iterator = self._stub.ReattachExecute(
+                                self._create_reattach_execute_request()
+                            )
+                            # shouldn't change
+                            assert not self._result_complete
+                            try:
+                                self._current = next(self._iterator)
+                            except StopIteration:
+                                pass
+                            has_next = self._current is not None
+                            if first_loop:
+                                # It's possible that the new iterator will be empty, so we need
+                                # to loop to get another. Eventually, there will be a non empty
+                                # iterator, because there's always a ResultComplete at the end
+                                # of the stream.
+                                first_loop = False
+                    return has_next
+            return False
+
+    def _release_execute(self, until_response_id: Optional[str]) -> None:
+        """
+        Inform the server to release the execution.
+
+        This will send an asynchronous RPC which will not block this iterator, the iterator can
+        continue to be consumed.
+
+        Release with untilResponseId informs the server that the iterator has been consumed until
+        and including response with that responseId, and these responses can be freed.
+
+        Release with None means that the responses have been completely consumed and informs the
+        server that the completed execution can be completely freed.
+        """
+        from pyspark.sql.connect.client.core import SparkConnectClient
+        from pyspark.sql.connect.client.core import Retrying
+
+        request = self._create_release_execute_request(until_response_id)
+
+        def target() -> None:
+            for attempt in Retrying(
+                can_retry=SparkConnectClient.retry_exception, **self._retry_policy
+            ):
+                with attempt:
+                    self._stub.ReleaseExecute(request)
+
+        threading.Thread(target=target).start()
```

Comment:
I can't find the symmetry way of doing StreamObserver in Python. I know this isn't super nice but it would at least work.

Reply:
This is fine. Maybe would have been even cleaner this way in scala, because the current code there of invoking an async execution, and then chaining another async execution from the onError of it is actually a bit funky.
A thread with retry loop over sync requests would be simpler code.